File: mini-diffusion-python/mini_diffusion/tensor.py
"""
Tensor operations for diffusion models.

Provides a simple wrapper around NumPy arrays with convenience methods
for 4D image tensors [batch, channels, height, width].
"""

import numpy as np
from typing import Union, Tuple, Optional


class Tensor:
    """4D tensor wrapper with diffusion-specific operations."""
    
    def __init__(self, data: np.ndarray):
        """Create tensor from numpy array."""
        self.data = data.astype(np.float64)
    
    @property
    def shape(self) -> Tuple[int, ...]:
        """Get tensor shape."""
        return self.data.shape
    
    @property
    def ndim(self) -> int:
        """Get number of dimensions."""
        return self.data.ndim
    
    def dim(self, i: int) -> int:
        """Get size of dimension i."""
        return self.data.shape[i]
    
    @property
    def size(self) -> int:
        """Get total number of elements."""
        return self.data.size
    
    # Factory methods
    @staticmethod
    def zeros(*shape: int) -> 'Tensor':
        """Create tensor filled with zeros."""
        return Tensor(np.zeros(shape))
    
    @staticmethod
    def ones(*shape: int) -> 'Tensor':
        """Create tensor filled with ones."""
        return Tensor(np.ones(shape))
    
    @staticmethod
    def randn(*shape: int, rng: Optional[np.random.Generator] = None) -> 'Tensor':
        """Create tensor with random normal values."""
        if rng is None:
            rng = np.random.default_rng()
        return Tensor(rng.standard_normal(shape))
    
    @staticmethod
    def from_shape(shape: Tuple[int, ...], value: float = 0.0) -> 'Tensor':
        """Create tensor with given shape filled with value."""
        return Tensor(np.full(shape, value))
    
    @staticmethod
    def xavier(fan_in: int, fan_out: int, rng: Optional[np.random.Generator] = None) -> 'Tensor':
        """Xavier/Glorot initialization."""
        if rng is None:
            rng = np.random.default_rng()
        std = np.sqrt(2.0 / (fan_in + fan_out))
        return Tensor(rng.standard_normal((fan_in, fan_out)) * std)
    
    @staticmethod
    def kaiming(fan_in: int, fan_out: int, rng: Optional[np.random.Generator] = None) -> 'Tensor':
        """Kaiming/He initialization."""
        if rng is None:
            rng = np.random.default_rng()
        std = np.sqrt(2.0 / fan_in)
        return Tensor(rng.standard_normal((fan_in, fan_out)) * std)
    
    # Arithmetic operations
    def add(self, other: Union['Tensor', float]) -> 'Tensor':
        """Element-wise addition."""
        if isinstance(other, Tensor):
            return Tensor(self.data + other.data)
        return Tensor(self.data + other)
    
    def sub(self, other: Union['Tensor', float]) -> 'Tensor':
        """Element-wise subtraction."""
        if isinstance(other, Tensor):
            return Tensor(self.data - other.data)
        return Tensor(self.data - other)
    
    def mul(self, other: Union['Tensor', float]) -> 'Tensor':
        """Element-wise multiplication."""
        if isinstance(other, Tensor):
            return Tensor(self.data * other.data)
        return Tensor(self.data * other)
    
    def div(self, other: Union['Tensor', float]) -> 'Tensor':
        """Element-wise division."""
        if isinstance(other, Tensor):
            return Tensor(self.data / other.data)
        return Tensor(self.data / other)
    
    def neg(self) -> 'Tensor':
        """Negate tensor."""
        return Tensor(-self.data)
    
    # Operator overloads
    def __add__(self, other: Union['Tensor', float]) -> 'Tensor':
        return self.add(other)
    
    def __radd__(self, other: float) -> 'Tensor':
        return self.add(other)
    
    def __sub__(self, other: Union['Tensor', float]) -> 'Tensor':
        return self.sub(other)
    
    def __rsub__(self, other: float) -> 'Tensor':
        return Tensor(other - self.data)
    
    def __mul__(self, other: Union['Tensor', float]) -> 'Tensor':
        return self.mul(other)
    
    def __rmul__(self, other: float) -> 'Tensor':
        return self.mul(other)
    
    def __truediv__(self, other: Union['Tensor', float]) -> 'Tensor':
        return self.div(other)
    
    def __neg__(self) -> 'Tensor':
        return self.neg()
    
    # Matrix operations
    def matmul(self, other: 'Tensor') -> 'Tensor':
        """Matrix multiplication."""
        return Tensor(self.data @ other.data)
    
    def __matmul__(self, other: 'Tensor') -> 'Tensor':
        return self.matmul(other)
    
    def transpose(self, *axes: int) -> 'Tensor':
        """Transpose tensor."""
        if len(axes) == 0:
            return Tensor(self.data.T)
        return Tensor(np.transpose(self.data, axes))
    
    @property
    def T(self) -> 'Tensor':
        """Shorthand for transpose."""
        return self.transpose()
    
    def reshape(self, *shape: int) -> 'Tensor':
        """Reshape tensor."""
        return Tensor(self.data.reshape(shape))
    
    def flatten(self) -> 'Tensor':
        """Flatten to 1D."""
        return Tensor(self.data.flatten())
    
    # Activation functions
    def relu(self) -> 'Tensor':
        """ReLU activation."""
        return Tensor(np.maximum(self.data, 0))
    
    def sigmoid(self) -> 'Tensor':
        """Sigmoid activation."""
        return Tensor(1.0 / (1.0 + np.exp(-self.data)))
    
    def tanh(self) -> 'Tensor':
        """Tanh activation."""
        return Tensor(np.tanh(self.data))
    
    def gelu(self) -> 'Tensor':
        """GELU activation (approximation)."""
        return Tensor(0.5 * self.data * (1 + np.tanh(
            np.sqrt(2 / np.pi) * (self.data + 0.044715 * self.data ** 3)
        )))
    
    def silu(self) -> 'Tensor':
        """SiLU/Swish activation."""
        return Tensor(self.data / (1.0 + np.exp(-self.data)))
    
    def softmax(self, axis: int = -1) -> 'Tensor':
        """Softmax along axis."""
        exp_x = np.exp(self.data - np.max(self.data, axis=axis, keepdims=True))
        return Tensor(exp_x / np.sum(exp_x, axis=axis, keepdims=True))
    
    # Statistics
    def mean(self, axis: Optional[int] = None) -> Union['Tensor', float]:
        """Compute mean."""
        result = np.mean(self.data, axis=axis)
        return float(result) if axis is None else Tensor(result)
    
    def std(self, axis: Optional[int] = None) -> Union['Tensor', float]:
        """Compute standard deviation."""
        result = np.std(self.data, axis=axis)
        return float(result) if axis is None else Tensor(result)
    
    def var(self, axis: Optional[int] = None) -> Union['Tensor', float]:
        """Compute variance."""
        result = np.var(self.data, axis=axis)
        return float(result) if axis is None else Tensor(result)
    
    def sum(self, axis: Optional[int] = None) -> Union['Tensor', float]:
        """Compute sum."""
        result = np.sum(self.data, axis=axis)
        return float(result) if axis is None else Tensor(result)
    
    def max(self, axis: Optional[int] = None) -> Union['Tensor', float]:
        """Compute max."""
        result = np.max(self.data, axis=axis)
        return float(result) if axis is None else Tensor(result)
    
    def min(self, axis: Optional[int] = None) -> Union['Tensor', float]:
        """Compute min."""
        result = np.min(self.data, axis=axis)
        return float(result) if axis is None else Tensor(result)
    
    # Indexing
    def __getitem__(self, idx):
        """Get item or slice."""
        result = self.data[idx]
        if isinstance(result, np.ndarray):
            return Tensor(result)
        return float(result)
    
    def __setitem__(self, idx, value):
        """Set item or slice."""
        if isinstance(value, Tensor):
            self.data[idx] = value.data
        else:
            self.data[idx] = value
    
    # Utility
    def clone(self) -> 'Tensor':
        """Create a copy of the tensor."""
        return Tensor(self.data.copy())
    
    def numpy(self) -> np.ndarray:
        """Get underlying numpy array."""
        return self.data
    
    def __repr__(self) -> str:
        return f"Tensor(shape={self.shape})"
    
    def __str__(self) -> str:
        return str(self.data)

File: mini-diffusion-java/src/main/java/com/minidiffusion/Tensor.java
package com.minidiffusion;

import java.util.Arrays;
import java.util.Random;
import java.util.function.DoubleUnaryOperator;

/**
 * A simple 4D tensor implementation for diffusion models.
 * Optimized for image data: [batch, channels, height, width]
 */
public class Tensor {
    private final double[] data;
    private final int[] shape;
    private final int[] strides;

    public Tensor(int... shape) {
        this.shape = shape.clone();
        this.strides = computeStrides(shape);
        int size = 1;
        for (int s : shape) size *= s;
        this.data = new double[size];
    }

    public Tensor(double[] data, int... shape) {
        this.shape = shape.clone();
        this.strides = computeStrides(shape);
        int size = 1;
        for (int s : shape) size *= s;
        if (data.length != size) {
            throw new IllegalArgumentException("Data length doesn't match shape");
        }
        this.data = data.clone();
    }

    private static int[] computeStrides(int[] shape) {
        int[] strides = new int[shape.length];
        int stride = 1;
        for (int i = shape.length - 1; i >= 0; i--) {
            strides[i] = stride;
            stride *= shape[i];
        }
        return strides;
    }

    // Factory methods
    public static Tensor zeros(int... shape) {
        return new Tensor(shape);
    }

    public static Tensor ones(int... shape) {
        Tensor t = new Tensor(shape);
        Arrays.fill(t.data, 1.0);
        return t;
    }

    public static Tensor randn(Random rng, int... shape) {
        Tensor t = new Tensor(shape);
        for (int i = 0; i < t.data.length; i++) {
            t.data[i] = rng.nextGaussian();
        }
        return t;
    }

    public static Tensor randUniform(Random rng, double low, double high, int... shape) {
        Tensor t = new Tensor(shape);
        for (int i = 0; i < t.data.length; i++) {
            t.data[i] = low + rng.nextDouble() * (high - low);
        }
        return t;
    }

    public static Tensor kaiming(Random rng, int fanIn, int fanOut) {
        double std = Math.sqrt(2.0 / fanIn);
        Tensor t = new Tensor(fanIn, fanOut);
        for (int i = 0; i < t.data.length; i++) {
            t.data[i] = rng.nextGaussian() * std;
        }
        return t;
    }

    // Getters
    public int[] getShape() { return shape.clone(); }
    public int dim(int i) { return shape[i]; }
    public int size() { return data.length; }

    // Element access
    private int index(int... indices) {
        int idx = 0;
        for (int i = 0; i < indices.length; i++) {
            idx += indices[i] * strides[i];
        }
        return idx;
    }

    public double get(int... indices) {
        return data[index(indices)];
    }

    public void set(double value, int... indices) {
        data[index(indices)] = value;
    }

    // Operations
    public Tensor clone() {
        return new Tensor(data.clone(), shape);
    }

    public Tensor add(Tensor other) {
        Tensor result = new Tensor(shape);
        for (int i = 0; i < data.length; i++) {
            result.data[i] = data[i] + other.data[i];
        }
        return result;
    }

    public Tensor sub(Tensor other) {
        Tensor result = new Tensor(shape);
        for (int i = 0; i < data.length; i++) {
            result.data[i] = data[i] - other.data[i];
        }
        return result;
    }

    public Tensor mul(Tensor other) {
        Tensor result = new Tensor(shape);
        for (int i = 0; i < data.length; i++) {
            result.data[i] = data[i] * other.data[i];
        }
        return result;
    }

    public Tensor scale(double s) {
        Tensor result = new Tensor(shape);
        for (int i = 0; i < data.length; i++) {
            result.data[i] = data[i] * s;
        }
        return result;
    }

    public Tensor addScalar(double s) {
        Tensor result = new Tensor(shape);
        for (int i = 0; i < data.length; i++) {
            result.data[i] = data[i] + s;
        }
        return result;
    }

    public Tensor apply(DoubleUnaryOperator op) {
        Tensor result = new Tensor(shape);
        for (int i = 0; i < data.length; i++) {
            result.data[i] = op.applyAsDouble(data[i]);
        }
        return result;
    }

    // Matrix multiply for 2D tensors
    public Tensor matmul(Tensor other) {
        if (shape.length != 2 || other.shape.length != 2) {
            throw new IllegalArgumentException("matmul requires 2D tensors");
        }
        int m = shape[0];
        int k = shape[1];
        int n = other.shape[1];
        if (k != other.shape[0]) {
            throw new IllegalArgumentException("Incompatible shapes for matmul");
        }
        
        Tensor result = new Tensor(m, n);
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                double sum = 0;
                for (int p = 0; p < k; p++) {
                    sum += get(i, p) * other.get(p, j);
                }
                result.set(sum, i, j);
            }
        }
        return result;
    }

    // Activations
    public Tensor relu() {
        return apply(x -> Math.max(0, x));
    }

    public Tensor sigmoid() {
        return apply(x -> 1.0 / (1.0 + Math.exp(-Math.max(-500, Math.min(500, x)))));
    }

    public Tensor silu() {
        return apply(x -> x / (1.0 + Math.exp(-Math.max(-500, Math.min(500, x)))));
    }

    public Tensor gelu() {
        return apply(x -> 0.5 * x * (1.0 + Math.tanh(Math.sqrt(2.0 / Math.PI) * (x + 0.044715 * x * x * x))));
    }

    // Statistics
    public double sum() {
        double s = 0;
        for (double v : data) s += v;
        return s;
    }

    public double mean() {
        return sum() / data.length;
    }

    public double variance() {
        double m = mean();
        double v = 0;
        for (double d : data) {
            double diff = d - m;
            v += diff * diff;
        }
        return v / data.length;
    }

    @Override
    public String toString() {
        return String.format("Tensor%s mean=%.4f", Arrays.toString(shape), mean());
    }
}
